internal_packages/keybase/lib/pgp-key-store.cjsx

```diff
@@ -0,0 +1,401 @@
+
+  _populate: (isPub) =>
+    # add metadata elements (sans keys) to later be populated with actual keys
+    # from disk
+    if isPub
+      keyDirectory = @_pubKeyDir
+      @_pubKeys = []
+    else
+      keyDirectory = @_privKeyDir
+      @_privKeys = []
+    fs.readdir(keyDirectory, (err, filenames) =>
+      i = 0
+      if filenames.length == 0
+        @trigger(@)
+      else
+        while i < filenames.length
+          filename = filenames[i]
+          if filename[0] == '.'
+            continue
+          absname = path.join(keyDirectory, filename)
+          key = {
+            path: absname,
+            addresses: filename.split(" ")
+          }
+          if isPub
+            @_pubKeys.push(key)
+          else
+            @_privKeys.push(key)
+          @trigger(@)
+          i++)
+
+  getKeyContents: ({key, passphrase}) =>
+    # Reads an actual PGP key from disk and adds it to the preexisting metadata
+    fs.readFile(key.path, (err, data) =>
+      pgp.KeyManager.import_from_armored_pgp {
+        armored: data
+      }, (err, km) =>
+        if err
+          # TODO wtf is that 'no header found' error
+          console.warn err
+        else
+          if km.is_pgp_locked()
+            if passphrase?
+              km.unlock_pgp { passphrase: passphrase }, (err) =>
+                if err
+                  console.warn err
+            else
+              console.error "No passphrase provided, but key is private."
+          timeout = 1000 * 60 * 30 # 30 minutes in ms
+          # NOTE this only allows for one priv key per address
+          # if it's already there, update, else insert
+          key.key = km
+          key.timeout = Date.now() + timeout
+          @getKeybaseData(key)
+        @trigger(@)
+    )
+
+  getKeybaseData: (key) =>
+    # Given a key, fetches metadata from keybase about that key
+    if not key.key?
+      PGPKeyStore.getKeyContents(key: key)
+    else
+      fingerprint = key.key.get_pgp_fingerprint().toString('hex')
+      kb.getUser(fingerprint, 'key_fingerprint', (err, user) =>
+        if user?.length == 1
+          key.keybase_user = user[0]
+          @trigger(@)
+      )
+
+  saveNewKey: (address, contents, isPub) =>
+    # Validate the email address, then write to file.
+    if @validAddress(address, isPub)
+      if isPub
+        keyPath = path.join(@_pubKeyDir, address)
+      else
+        keyPath = path.join(@_privKeyDir, address)
+      # Just say no to trailing whitespace.
+      if contents.charAt(contents.length - 1) != '-'
+        contents = contents.slice(0, -1)
+      fs.writeFile(keyPath, contents, (err) =>
+        if (err)
+          @_displayError(err)
+      )
+
+  deleteKey: (key) =>
+    if this._displayDialog(
+      'Delete this key?',
+      'The key will be permanently deleted.',
+      ['Delete', 'Cancel']
+    )
+      fs.unlink(key.path, (err) =>
+        if (err)
+          @_displayError(err)
+      )
+
+  addAddressToKey: (profile, address) =>
```

Comment:
TODO: calling this function will cause a duplicate key to be added (new file created and subsequently cached, old key not removed from the cache) that makes React very upset due to duplicate `key` props when it goes to render.